### PR TITLE
Finish the work on `try_apply`

### LINKED ISSUE #6182: Non-panicking Version of Reflect.apply()
## What problem does this solve or what need does it fill?

While working on a reflect-based scripting solution, I've run into the need to try and `apply()` one reflect value to another, but I can't have it panic if it fails, I need to return an error instead.

## What solution would you like?

It would be nice to have maybe a `try_apply` method or just return a result from the existing `apply` method instead of panicking.

## What alternative(s) have you considered?

I tried to use catch_unwind to capture the panic, but I'm working with types that are not unwind safe, so it's not an option for me in this case.

I believe I'll have to compare the values that I'm applying manually, with a match on their ReflectRef's in order to see if they are compatible, before doing the apply.

---

Another alternative would be to provide an `is_compatible()` method or something like that, that's essentially just a check on whether or not an `apply` would succeed, but I'm not sure if that's more useful or not. I still feel like panicking by default here isn't the best option.

## Additional context

None yet.

### PR DESCRIPTION
**Pull request by Brezak**
# Objective

Finish the `try_apply` implementation started in #6770 by @feyokorenhof.
Supersedes and closes #6770. Closes #6182

## Solution

Add `try_apply` to `Reflect` and implement it in all the places that implement `Reflect`.

---

## Changelog

Added `try_apply` to `Reflect`.

### CODE REVIEW

**Comment by Brezak** [comment]
Current TODO:

- [ ] Figure out how to properly handle errors
    - [ ] Figure out which fields of the enum can be turned into static strings.
    - [ ] Maybe switch to a opaque struct with a kinds function.
- [ ] Write tests using try_apply (apply defaults to using try_apply but we should test if the proper errors are returned).
- [ ] Write an example.

**Comment by Brezak** [comment]
The code is mostly finished. I'm just not confident about `ApplyError`. It's sometimes not very clear about where exactly the mistake occurred. It would help if `FromReflect` also returned an error. Then we could unify their error types under a `ReflectError`, but this would be a breaking change and I don't know if we want to leave that for a future PR.

File: crates/bevy_reflect/bevy_reflect_derive/src/enum_utility.rs [hunk 1]
```diff
@@ -50,17 +50,23 @@ pub(crate) fn get_variant_constructors(
...
                     };
                     let ty = field.data.ty.to_token_stream();
 
-                    let on_error = format!("{field_ref_str} should be of type `{ty}`");
-                    let on_missing = format!("{field_ref_str} is required but could not be found");
-
-                    (quote!(.expect(#on_error)), quote!(.expect(#on_missing)))
+                    (
+                        quote!(.ok_or(#bevy_reflect_path::ApplyError::MismatchedTypes(
+                            #field_ref_str.into(),
```
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/enum_utility.rs:62]: Ideally, `into` calls would also be fully qualified.

**Review by MrGVSV — COMMENTED** [review]
Looks good! Just a few comments/suggestions.

I also added suggestions to move some of the macro code to use fully-qualified syntax (hopefully with the bulk-apply for PR suggestions, that can take some of the work off your plate).

File: crates/bevy_reflect/src/impls/std.rs [hunk 2]
```diff
@@ -1056,13 +1073,18 @@ impl Reflect for Cow<'static, str> {
...
+    fn try_apply(&mut self, value: &dyn Reflect) -> Result<(), ApplyError> {
+        let any = value.as_any();
+        if let Some(value) = any.downcast_ref::<Self>() {
             *self = value.clone();
         } else {
-            panic!("Value is not a {}.", Self::type_path());
+            return Err(ApplyError::MismatchedTypes(
+                value.reflect_type_path().into(),
+                // If we invoke the reflect_type_path on self directly the borrow checker complains that the lifetime of self must outlive 'static
+                <Self as DynamicTypePath>::reflect_type_path(self).into(),
```
[the remark this answers is not in the document]
Comment (Brezak) [reply crates/bevy_reflect/src/impls/std.rs]: No idea. Good catch.

File: crates/bevy_reflect/src/utility.rs [hunk 3]
```diff
@@ -71,19 +71,19 @@ mod sealed {
...
+/// #   fn as_any_mut(&mut self) -> &mut dyn Any { todo!() }
+/// #   fn into_reflect(self: Box<Self>) -> Box<dyn Reflect> { todo!() }
+/// #   fn as_reflect(&self) -> &dyn Reflect { todo!() }
+/// #   fn as_reflect_mut(&mut self) -> &mut dyn Reflect { todo!() }
+/// #   fn try_apply(&mut self, value: &dyn Reflect) -> Result<(), ApplyError> { todo!() }
+/// #   fn set(&mut self, value: Box<dyn Reflect>) -> Result<(), Box<dyn Reflect>> { todo!() }
+/// #   fn reflect_ref(&self) -> ReflectRef { todo!() }
+/// #   fn reflect_mut(&mut self) -> ReflectMut { todo!() }
+/// #   fn reflect_owned(self: Box<Self>) -> ReflectOwned { todo!() }
+/// #   fn clone_value(&self) -> Box<dyn Reflect> { todo!() }
```
Comment (soqb) [inline crates/bevy_reflect/src/utility.rs:86]: hmm. looks like formatting degraded here:

```suggestion
/// #     fn get_represented_type_info(&self) -> Option<&'static TypeInfo> { todo!() }
/// #     fn into_any(self: Box<Self>) -> Box<dyn Any> { todo!() }
/// #     fn as_any(&self) -> &dyn Any { todo!() }
/// #     fn as_any_mut(&mut self) -> &mut dyn Any { todo!() }
/// #     fn into_reflect(self: Box<Self>) -> Box<dyn Reflect> { todo!() }
/// #     fn as_reflect(&self) -> &dyn Reflect { todo!() }
/// #     fn as_reflect_mut(&mut self) -> &mut dyn Reflect { todo!() }
/// #     fn try_apply(&mut self, value: &dyn Reflect) -> Result<(), ApplyError> { todo!() }
/// #     fn set(&mut self, value: Box<dyn Reflect>) -> Result<(), Box<dyn Reflect>> { todo!() }
/// #     fn reflect_ref(&self) -> ReflectRef { todo!() }
/// #     fn reflect_mut(&mut self) -> ReflectMut { todo!() }
/// #     fn reflect_owned(self: Box<Self>) -> ReflectOwned { todo!() }
/// #     fn clone_value(&self) -> Box<dyn Reflect> { todo!() }
```

File: crates/bevy_reflect/src/utility.rs [hunk 4]
```diff
@@ -150,19 +150,19 @@ impl<T: TypedProperty> Default for NonGenericTypeCell<T> {
...
```
Comment (soqb) [inline crates/bevy_reflect/src/utility.rs:165]: same here:

```suggestion
/// #     fn get_represented_type_info(&self) -> Option<&'static TypeInfo> { todo!() }
/// #     fn into_any(self: Box<Self>) -> Box<dyn Any> { todo!() }
/// #     fn as_any(&self) -> &dyn Any { todo!() }
/// #     fn as_any_mut(&mut self) -> &mut dyn Any { todo!() }
/// #     fn into_reflect(self: Box<Self>) -> Box<dyn Reflect> { todo!() }
/// #     fn as_reflect(&self) -> &dyn Reflect { todo!() }
/// #     fn as_reflect_mut(&mut self) -> &mut dyn Reflect { todo!() }
/// #     fn try_apply(&mut self, value: &dyn Reflect) -> Result<(), ApplyError> { todo!() }
/// #     fn set(&mut self, value: Box<dyn Reflect>) -> Result<(), Box<dyn Reflect>> { todo!() }
/// #     fn reflect_ref(&self) -> ReflectRef { todo!() }
/// #     fn reflect_mut(&mut self) -> ReflectMut { todo!() }
/// #     fn reflect_owned(self: Box<Self>) -> ReflectOwned { todo!() }
/// #     fn clone_value(&self) -> Box<dyn Reflect> { todo!() }
```

File: crates/bevy_utils/README.md
Comment (soqb) [inline crates/bevy_utils/README.md:1]: why was this removed?

**Review by soqb — COMMENTED** [review]
all this LGTM. just two little things before i approve.

File: crates/bevy_utils/README.md
Comment (Brezak) [reply crates/bevy_utils/README.md:1]: `jj` was being fucky. I've restored the file.

Comment (soqb) [reply crates/bevy_utils/README.md:1]: i await the day we all switch to [Pijul](https://pijul.org/).

**Review by soqb — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]

**Pull request merged**